src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+        /// </summary>
+        private async ValueTask HightlightNodeAsync(CancellationToken cancellationToken)
+        {
+            await TaskScheduler.Default;
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null)
+                {
+                    var documentSymbolModelArray = ((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource).ToImmutableArray();
+                    var caretPoint = activeTextView.GetCaretPoint(activeTextView.TextBuffer);
+                    if (caretPoint.HasValue)
+                    {
+                        var caretPosition = caretPoint.Value.Position;
+                        DocumentOutlineHelper.SelectDocumentNode(documentSymbolModelArray, activeTextView.TextSnapshot, LspSnapshot, caretPosition);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Starts a new task to get the current document model.
+        /// </summary>
+        private void StartGetModelTask()
+        {
+            _getModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to update the current document model.
+        /// </summary>
+        private void StartModelUpdateTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _updateModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to highlight the symbol node corresponding to the current caret position in the editor.
+        /// </summary>
+        private void StartHightlightNodeTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _highlightNodeQueue.AddWork();
+        }
+
+        private void ExpandAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, true);
+        }
+
+        private void CollapseAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, false);
+        }
+
+        private void Search(object sender, EventArgs e)
+        {
+            StartModelUpdateTask();
+        }
+
+        private void SortByName(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Name;
+            StartModelUpdateTask();
+        }
+
+        private void SortByOrder(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Order;
+            StartModelUpdateTask();
+        }
+
+        private void SortByType(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Type;
+            StartModelUpdateTask();
+        }
+
+        // When symbol node clicked, select the corresponding code
```

Comment:
Consider XML doc comments here to make it easier to see what this does when browsing the code.